torch/distributed/rpc/api.py

```diff
@@ -210,7 +210,15 @@ def _all_gather(obj, worker_names=None, timeout=UNSET_RPC_TIMEOUT):
 
     with _all_gather_dict_lock:
         states = _all_gather_sequence_id_to_states[sequence_id]
-    states.proceed_signal.wait()
+
+    timed_out = not states.proceed_signal.wait(timeout=timeout)
```

Comment:
Imagine a case where there is a primary node doing RPCs to workers. It would have code like this:

```
rpc.init(...)
for i in range(100000):
  rpc.async(...)
rpc.shutdown(...)
```

The workers would have code like this:

```
rpc.init(...)
rpc.shutdown(...)
```

Now since the default shutdown timeout is just 5seconds, the workers will just time out and die before the primary worker has completed. The shutdown sequence should only start when all RPC workers enter the `rpc.shutdown` call and this is what the `wait()` with no timeout provides.

Reply:
Regarding the test failures, is this happening because a process in the test is stuck and never reaches shutdown? If so, I think the proper fix is that we should determine why it is stuck and resolve that issue.

Reply:
Makes sense, and yes the test failures is happening because a process gets stuck here. 

Agreed that we should determine why it is stuck, but right now the process just gets stuck and doesn't print out any logs, it was hard to trace down to it getting stuck here, hence the thinking that a timeout could help.